tests/unit/accounts/test_services.py

```diff
@@ -886,6 +887,107 @@ def test_check_recovery_code(self, user_service, metrics):
             ),
         ]
 
+    def test_generate_device_id_secret(self, user_service, monkeypatch):
+        user = UserFactory.create()
+
+        device_id_secret1 = user_service.generate_device_id_secret(user.id)
+        assert device_id_secret1 is not None
+        assert isinstance(device_id_secret1, str)
+        assert len(device_id_secret1) >= 32
+
+        # Make sure the same user can generate a 2nd device_id_secret
+        device_id_secret2 = user_service.generate_device_id_secret(user.id)
+        assert device_id_secret2 is not None
+        assert isinstance(device_id_secret2, str)
+        assert len(device_id_secret2) >= 32
+        assert device_id_secret1 != device_id_secret2
+
+        # Make sure the code handles things properly when it cannot generate an unused
+        # secret
+        monkeypatch.setattr(
+            secrets, "token_urlsafe", lambda length: "same_thing_every_time"
+        )
+        user_service.generate_device_id_secret(user.id)
+        with pytest.raises(RuntimeError):
+            user_service.generate_device_id_secret(user.id)
+
+    def test_check_device_valid(self, user_service, metrics):
```

Comment:
Breaking this test up into separate cases would be beneficial. That's in particular to the assertions around metrics, but overall its good practice.